### DISCUSSION TITLE
Add asyncio support?

### DISCUSSION THREAD

**Discussion by igo95862** [pyudev/pyudev#483] [Show and tell]
Currently the only ways to monitor the devices are synchronous: https://pyudev.readthedocs.io/en/latest/guide.html#asynchronous-monitoring

Asyncio was added to Python 3.5 and does not work well with long sync calls.

Asyncio supports watching a file descriptor: https://docs.python.org/3/library/asyncio-eventloop.html#watching-file-descriptors

It should be possible to integrate that with `udev_monitor_get_fd` to watch a file descriptor.

**Comment by garyvdm** [comment]
I wrote some simple code to monitor devices with asyncio. Posting it here incase it's useful:

```
from asyncio import Event, get_event_loop
from typing import AsyncGenerator

from pyudev import Context, Device, Monitor


async def iter_monitor_devices(context: Context, **kwargs) -> AsyncGenerator[Device, None]:
    for device in context.list_devices(**kwargs):
        yield device

    monitor = Monitor.from_netlink(context)
    monitor.filter_by(**kwargs)
    monitor.start()
    fd = monitor.fileno()
    read_event = Event()
    loop = get_event_loop()
    loop.add_reader(fd, read_event.set)
    try:
        while True:
            await read_event.wait()
            while True:
                device = monitor.poll(0)
                if device:
                    yield device
                else:
                    read_event.clear()
                    break
    finally:
        loop.remove_reader(fd)
```

Example:

```
from asyncio import run

async def monitor_usb()
    context = Context()
    async for device in iter_monitor_devices(context, subsystem='usb'):
        print(device)

run(monitor_usb())
```

**Reply by bidasknakayama** [reply]
This was very useful. Thank you !!!
Please add into official module too.

Kazu

**Comment by askolo4ek** [comment]
@garyvdm Thanks for this code. It's just what I need at work <3
